File: PYTHON/Shell_Sort.py
def shellSort(data,n):
    h = n//2
    while h > 0:
        shell(data,h,n)
        h = h//2
    
    return data
def shell(data,h,n):
    index = 0
    while (index+h) <= 0:
        index += h
    while index > 0 :
        indexData1 = 0
        indexData2 = indexData1+h
        while indexData2 < index :
            if data[indexData1] > data[indexData2] :
                temp = data[indexData1]
                data[indexData1],data[indexData2] = data[indexData2],temp
            indexData1 = indexData2
            indexData2 += h
        index = index-h   

data = list(map(int,input('Enter Input : ').split(' ')))

print(shellSort(data,len(data)))

File: JAVA/Shell Sort.java
import java.util.Scanner;

public class ShellSort {
	int shellsort(int arr[], int n) {
		//here n is the size of the array
		//initially gap = n/2
		for (int gap = n / 2; gap > 0; gap = gap / 2) {
			for (int i = gap; i < n; i++) {
				//decreasing the gap
				int k = arr[i];
				int j = i;

				while (j >= gap && arr[j - gap] > k) {
					arr[j] = arr[j - gap];
					j = j - gap;
				}
				arr[j] = k;
			}
		}
		return 0;
	}

	private static Scanner sc = new Scanner(System.in);

	public static void main(String args[]) {
		//we store the number of elements in n
		System.out.println("Kindly enter the number of elements: ");
		int n = sc.nextInt();

		//declaring an array of n elements
		int array[] = new int[n];

		//taking the input from the user
		for (int i = 0; i < n; i++) {
			System.out.print("\n" + i + "'th element : ");
			array[i] = sc.nextInt();
		}

		System.out.println("\nArray before sorting");
		for (int j = 0; j < n; j++)
			System.out.print(array[j] + " ");


		//creating an object of shellsort class so that the methods inside shellsort class can be used
		ShellSort object = new ShellSort();
		object.shellsort(array, n);

		System.out.println("\nArray after sorting");
		for (int k = 0; k < n; k++)
			System.out.print(array[k] + " ");

	}
}
